### ISSUE TITLE
The crypto-transforms window retains key information even after you close the file

### ISSUE THREAD

**Issue by karagog** [karagog/Gryptonite#3]
If you open crypto-transoforms when a file is open, it draws the key information from the current session. But when you close the file, the key is still set in the window. A malicious user cannot get the key (short of randomly guessing it) but they could use the key to decrypt something that you generated with that window.

Note that they can't use this key to decrypt your password database, because the database key requires salt for the key derivation function.

**Comment by karagog** [comment]
I just realized that it shouldn't get the key information from the database session in any case, because if they encrypt something using one database's key information, they will not be able to decrypt it with a different database's key info, even if the two databases use the same password. The abstraction I want to provide with this window is that just their password decrypts the file.

**Comment by karagog** [comment]
In order to achieve this securely, when you encrypt each file you have to prepend the key-salt to it. So every crypt-transforms session will have a salt associated with it, and everything the user encrypts will have the salt in front, followed by the payload and nonce.

**Comment by karagog** [comment]
I can't add salt to the crypto-transforms window, because after you set the key you should be able to decrypt anything that was encrypted using the same password, even if it was encrypted in a different session (which would be with a different salt). So the crypto-transforms encrypt/decrypt methods will be inherently weaker than the rest of the keyfile, because it doesn't use salt in the Key Derivation Function. That's okay for me (for now) because it's really only intended to be a curiosity not an "official" encryption solution.

**Comment by karagog** [comment]
I guess you COULD make it work by having the crypto transforms window remember the password instead of the key. then it could dynamically generate the proper key when it tries to decrypt the file (with salt prepended). My only problem with this is that I wanted the interface to never have your master password sitting around in memory. In some cases it's unavoidable (like the GetPasswordDialog, for example, will have it stored in the line edit until it's deleted). But in all cases I have control over, I always throw away the password once I generate the key from it. I'll have to think about this more. In the meantime, just make sure the window starts without key information.

**Comment by karagog** [comment]
I just fixed this by forcing the user to input the key information every time the crypto-transforms window is opened. That way the key is not tied to the file in any way (unless the user happens to use the same password for both, but there is no way to tell that).

I'm still wondering if I should close the crypto-transforms window when you close a file, but my gut instinct says I can leave it open, because really it's its own tool separate from the file.

Fixed in: 955efb9f8b92684e07e2d59842cfeb59b314ff06
